bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java

```diff
@@ -172,6 +174,14 @@ private void registerFirmwareUpdateHandler(Thing thing, ThingHandler thingHandle
         return serviceRegistration;
     }
 
+    protected ServiceRegistration<?> registerActionService(Thing thing, ThingHandler handler,
```

Comment:
I remember we discussed offline that ThingHandlers should provide a list of services (list of class objects) which then would be instantiated by the BaseThingHandlerFactory. The handler would then be injected by `setHandler` which we discussed should be defined on an extra interface. This interface must be provided by every class instance the handler names to be instantiated.
This solution could then be extended to automatically register discovery service.

Reply:
The BaseThingHandlerFactory would then also track the serviceRegs and unregister them again on handler removal.

Reply:
@kaikreuzer please add details which may have been lost in my memory. tnx.

Reply:
I proposed a solution to this in my newest commit.

Reply:
The proposal looks good! This can then be removed.